File: app/src/main/java/com/ignis/android_cleanarchitecture/presentation/view/adapter/ForecastRecyclerAdapter.java
package com.ignis.android_cleanarchitecture.presentation.view.adapter;

import android.content.Context;
import android.databinding.DataBindingUtil;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.ignis.android_cleanarchitecture.R;
import com.ignis.android_cleanarchitecture.databinding.RecyclerForecastBinding;
import com.ignis.android_cleanarchitecture.presentation.listener.view.adapter.ForecastRecyclerAdapterListener;
import com.ignis.android_cleanarchitecture.presentation.presenter.adapter.ForecastViewModel;

import java.util.ArrayList;
import java.util.List;

/**
 * Forecast RecyclerView Adapter
 * <p>
 * Created by tamura_k on 2016/05/31.
 */
public class ForecastRecyclerAdapter extends RecyclerView.Adapter<ForecastRecyclerAdapter.ViewHolder> {

    public static class ViewHolder extends RecyclerView.ViewHolder {

        private RecyclerForecastBinding binding;

        public ViewHolder(View view) {
            super(view);
            binding = DataBindingUtil.bind(itemView);
        }

        public RecyclerForecastBinding getBinding() {
            return binding;
        }

    }

    private Context context;
    private List<ForecastViewModel> forecastViewModelList;
    private ForecastRecyclerAdapterListener forecastRecyclerAdapterListener;

    public ForecastRecyclerAdapter(Context context) {
        this.context = context;
        this.forecastViewModelList = new ArrayList<>();
    }

    @Override
    public int getItemCount() {
        return forecastViewModelList.size();
    }

    @Override
    public ForecastRecyclerAdapter.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(context).inflate(R.layout.recycler_forecast, parent, false);
        return new ForecastRecyclerAdapter.ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(ForecastRecyclerAdapter.ViewHolder holder, int position) {
        final ForecastViewModel item = forecastViewModelList.get(position);

        holder.getBinding().setVariable(com.ignis.android_cleanarchitecture.BR.viewModel, item);
        holder.getBinding().executePendingBindings();
        holder.getBinding().getRoot().setOnClickListener(v -> {
            if (forecastRecyclerAdapterListener != null) forecastRecyclerAdapterListener.onItemClick(item);
        });
        item.loadImage(holder.getBinding().weatherImage);
    }

    public void setForecastViewModelList(List<ForecastViewModel> forecastViewModelList) {
        this.forecastViewModelList = forecastViewModelList;
    }

    public void setForecastRecyclerAdapterListener(ForecastRecyclerAdapterListener forecastRecyclerAdapterListener) {
        this.forecastRecyclerAdapterListener = forecastRecyclerAdapterListener;
    }

}

File: presentation/src/main/java/com/kazakago/cleanarchitecture/presentation/view/adapter/ForecastRecyclerAdapter.kt
package com.kazakago.cleanarchitecture.presentation.view.adapter

import android.content.Context
import android.support.v7.widget.RecyclerView
import android.view.ViewGroup
import com.kazakago.cleanarchitecture.domain.model.weather.Forecast
import com.kazakago.cleanarchitecture.presentation.R
import com.kazakago.cleanarchitecture.presentation.extension.formattedDateText
import com.squareup.picasso.Picasso
import kotlinx.android.synthetic.main.recycler_forecast.view.*

class ForecastRecyclerAdapter(private val context: Context) : RecyclerView.Adapter<ForecastRecyclerAdapter.ViewHolder>() {

    interface Listener {
        fun onItemClick(forecast: Forecast)
    }

    var listener: Listener? = null
    var forecastList: List<Forecast> = listOf()

    override fun getItemCount(): Int {
        return forecastList.size
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        return ViewHolder(context, parent)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        holder.item = forecastList[position]
    }

    inner class ViewHolder(context: Context, parent: ViewGroup) : AbsItemViewHolder<Forecast>(context, parent, R.layout.recycler_forecast) {
        init {
            itemView.setOnClickListener {
                item?.let { listener?.onItemClick(it) }
            }
        }
        override fun onSetItem(item: Forecast?) {
            Picasso.with(context).load(item?.imageUrl.toString())
                    .into(itemView.weatherImageView)
            itemView.dateLabelTextView.text = item?.dateLabel
            itemView.dateTextView.text = item?.date?.formattedDateText(context)
            itemView.telopTextView.text = item?.telop
            itemView.maxTemperatureTextView.text = context.getString(R.string.temperature_max, item?.maxTemperature?.toString() ?: "--")
            itemView.minTemperatureTextView.text = context.getString(R.string.temperature_min, item?.minTemperature?.toString() ?: "--")
        }
    }

}
